[Feature] Add documentation for "Throttle policy import-export support for APIM (#6389)
wso2/docs-apim: en/docs/design/rate-limiting/throttle-policy-import-export.md

# Importing and Exporting Throttle Policies

> Throttling policies exist at different levels such as Application Level, Subscriber Level, Application Level, Advanced Policy Level, etc. These throttling policies can belong to different environments according to the product lifecycle and are related to different APIs or Applications. Of these throttling policies, some of them exist as default policies belonging to every environment, and throttling policies can be custom-made. Users might need to import custom-made or default throttling policies from one environment to another.

Reviewer: Application Level is mentioned twice.

Reviewer 2: ```suggestion
Throttling policies exist at different levels such as Application Level, Subscriber Level, Advanced Policy Level, etc. These throttling policies can belong to different environments according to the product lifecycle and are related to different APIs or Applications. Of these throttling policies, some of them exist as default policies belonging to every environment, and throttling policies can be custom-made. You may need to import custom-made or default throttling policies from one environment to another.
```

> When an API or an Application is imported from an environment that has custom-made throttling policies to another environment, if the API or the  Application has subscribed to a throttling policy that is not in the destination environment, an error is thrown indicating the throttling policy is not supported and the importation fails.

Reviewer: Good to reword this. Instead of saying an error is thrown, we can say `... if the API or the  Application has subscribed to a throttling policy that is not in the destination environment, importation fails indicating the throttling policy is not supported.`

Reviewer 2: ```suggestion
You may encounter an error when an API or an Application is imported from an environment that has custom-made throttling policies to another environment. This happens if the API or the Application has subscribed to a throttling policy that is not in the destination environment. Importation fails to indicate the throttling policy is not supported.
```

> Before importing either the API or App to the destination environment,  throttling policies are imported to the destination environment using a separate step. After throttling policies are imported to the destination environment we can import the API/Application to the destination environment without any throttling policy support errors.

Reviewer 2: ```suggestion
Before importing either the API or App to the destination environment, throttling policies are imported to the destination environment using a separate step. After throttling policies are imported to the destination environment we can import the API/Application to the destination environment without any throttling policy support errors.
```

## Operations

This feature introduces 3 new operations

1. Get throttling policies
2. Export throttling policy
3. Import throttling policy

Alongside those new operations, 3 new Admin APIs and 3 new APICTL commands are introduced.

### Admin APIs

1. Unified Search

    ```
    GET /throttling/policies/search

    Retrieve/Search Throttling Policies.
    ```

> 2. Import Export Throttling Policies

Reviewer 2: ```suggestion
2. Import/Export Throttling Policies
```

| custom        | to export only custom throttling policies        |
| subscription  | to export only subscription throttling policies  |
| application   | to export only application throttling policies   |

`-- format` - File format of exported file(json or yaml) (default "YAML")

``` tab="Example"
export  policy  rate-limiting   -n Silver  -e prod   -- type subscription
```

!!! info "Exported File"
    ```
    type: throttling policy
    subtype: subscription policy
    version: v4.1.0
    data:
    policyId: 15797f4c-3742-4909-920c-13f164fd2b1c
    policyName: Silver
    displayName: Silver
    description: Allows 2000 requests per minute
    isDeployed: true
    type: SubscriptionThrottlePolicy
    graphQLMaxComplexity: 0
    graphQLMaxDepth: 0
    defaultLimit:
        type: REQUESTCOUNTLIMIT
        requestCount:
        timeUnit: min
        unitTime: 1
        requestCount: 2000
        bandwidth: null
        eventCount: null
    monetization: null
    rateLimitCount: 0
    rateLimitTimeUnit: null
    subscriberCount: 0
    customAttributes: []
    stopOnQuotaReach: true
    billingPlan: FREE
    permissions: null
    ```

### 3. Import Throttling Policy

> Import throttling policy operation allows users to import exported throttling policy files to a different environment and it also allows overwriting capability.

Reviewer 2: ```suggestion
Import throttling policy operation allows users to import exported throttling policy files to a different environment and it also has the capability to overwrite existing policies.
```
